### PR TITLE
Mul fix and fix #117

### PR DESCRIPTION
**Pull request by liliatangxy**
Enables proving for mul when we need to broadcast in the last dimension in vae-decoder. Also fixes some bugs in Transpose and Gather introduced in #117

### CODE REVIEW

File: src/layer/gather.rs [hunk 1]
```diff
@@ -32,15 +32,18 @@ impl Layer for GatherLayer {
...
-      if x > Fr::from(input_shapes[0][0] as i64) {
-        Fr::from(input_shapes[0][0] as i64) + x
-      } else {
-        x
-      }
-    });
-    let indices = graph.addBB(Box::new(Const2BasicBlock { c: indices }));
-    let indices_output = graph.addNode(indices, vec![]);
+    let mut indices_output = -2;
+    if input_shapes[1].len() == 0 {
```
Comment (ddkang) [inline src/layer/gather.rs:37]: What is going on here?

File: src/layer/transpose.rs [hunk 2]
```diff
@@ -23,14 +23,16 @@ impl Layer for TransposeLayer {
 
     if *axes.last().unwrap() == n - 1 {
       let transpose = graph.addBB(Box::new(TransposeBasicBlock { perm: axes.clone() }));
+      println!("axes: {:?}", axes);
```
Comment (ddkang) [inline src/layer/transpose.rs]: ?

File: src/layer/transpose.rs [hunk 3]
```diff
@@ -23,14 +23,16 @@ impl Layer for TransposeLayer {
...
       graph.outputs.push((output, 0));
     } else {
-      // move last element to be 2nd to last
       let pos = axes.iter().position(|&x| x == n - 1).unwrap();
       let mut perm = axes.clone();
+      // keep n-1 in last index and move axes[n-1] to n-2
       perm[pos] = axes[n - 2];
-      perm[n - 2] = n - 1;
+      perm[n - 2] = axes[n - 1];
+      perm[n - 1] = n - 1;
```
Comment (ddkang) [inline src/layer/transpose.rs:42]: Is this always right?

File: src/layer/gather.rs (hunk 1 above)
Comment (liliatangxy) [reply src/layer/gather.rs:37]: Avoids performing unwrap on None for constant[1], added comment

File: src/layer/transpose.rs (hunk 3 above)
Comment (liliatangxy) [reply src/layer/transpose.rs:42]: Added comments about correctness

**Comment by Vimalakirti** [comment]
Why does the build fail?

**Comment by liliatangxy** [comment]
Seems like there is a package-related error with using ndarray v0.15.6 and v.0.16.1. I can't replicate it on the cc machines or my mac though with the most recent nightly build. Here's the package tree I see on mac.

```
ndarray v0.15.6
├── tract-core v0.21.6
│   ├── tract-hir v0.21.6
│   │   └── tract-onnx v0.21.6
│   │       └── zk_torch v0.1.0 (/Users/lilia/workspace/zk-torch)
│   └── tract-nnef v0.21.6
│       ├── tract-onnx v0.21.6 (*)
│       └── tract-onnx-opl v0.21.6
│           └── tract-onnx v0.21.6 (*)
├── tract-data v0.21.6
│   ├── tract-core v0.21.6 (*)
│   └── tract-linalg v0.21.6
│       └── tract-core v0.21.6 (*)
└── zk_torch v0.1.0 (/Users/lilia/workspace/zk-torch)
```

**Comment by liliatangxy** [comment]
tract-core 0.21.7 is being used here which contains ndarray 0.16

**Comment by Vimalakirti** [comment]
> Seems like there is a package-related error with using ndarray v0.15.6 and v.0.16.1. I can't replicate it on the cc machines or my mac though with the most recent nightly build. Here's the package tree I see on mac.
> 
> ```
> ndarray v0.15.6
> ├── tract-core v0.21.6
> │   ├── tract-hir v0.21.6
> │   │   └── tract-onnx v0.21.6
> │   │       └── zk_torch v0.1.0 (/Users/lilia/workspace/zk-torch)
> │   └── tract-nnef v0.21.6
> │       ├── tract-onnx v0.21.6 (*)
> │       └── tract-onnx-opl v0.21.6
> │           └── tract-onnx v0.21.6 (*)
> ├── tract-data v0.21.6
> │   ├── tract-core v0.21.6 (*)
> │   └── tract-linalg v0.21.6
> │       └── tract-core v0.21.6 (*)
> └── zk_torch v0.1.0 (/Users/lilia/workspace/zk-torch)
> ```

Can we set `tract-onnx = "0.21.2"` back? I think it's not related to cc. The cpu build fails here.

**Comment by liliatangxy** [comment]
Sorry, that was a test.

**Comment by liliatangxy** [comment]
tract-onnx 0.21.2 is no longer available on crates.io, so we have to force a different version to prevent 0.21.7 from being used. The highest version which seems to work is 0.21.6 so I can force that version

**Review by Vimalakirti — APPROVED** [review]

**Review by ddkang — APPROVED** [review]

**Pull request merged**